writing/2019-10-05-Haskell-in-Production.md

```diff
@@ -0,0 +1,113 @@
+
+## Existing parts in this series
+
+* Part 0: Introduction (this page!)
+* Part 1: Making components testable
+<!-- * Part 2: Testing your components -->
+<!-- * Part 3: Deploying your application -->
+
+This section will be updated as more articles are published.
+
+## Companion repo
+For this series you can find the code that goes along with each part in the
+[haskell-in-production](https://github.com/felixmulder/haskell-in-production)
+repo on GitHub.
+
+## Concerns when writing a production service
+We have a couple of concerns that we would have no-matter which language we're
+writing our service in.
+
+* Multiple environments and regions
+* Testability
+* Performance & scalability
+* Maintainability
+
+When you're coming from object-oriented programming, you usually think of these
+things in order to facilitate the above:
+
+* Dependency injection
+* Mocking
+
+In functional programming we can do both, but the way we do it is much simpler
+than the tooling usually required by OO frameworks like Spring and Dropwizard.
+
+Below, I'll elaborate some more on the requirements posited above. If you want
+to skip ahead to how we'll solve these concerns - please see [part
+2](./2019/10/05-Haskell-in-Production.html) of this article series.
+
+### Multiple environments
+Klarna provides multiple environments for its services, we have (at least) the
+following:
+
+* Staging
+
+  This environment doesn't really have an SLA, but you should expect other
+  teams to write end-to-end tests against services that you deploy here.
+
+* Performance
+
+  This environment is used in order to do load testing, or stress tests on your
+  service
+
+* Production
+
+  Perhaps obvious, but this is the service that actually serves live traffic.
+  Depending on which team you're in you can have very strict requirements in
+  your SLA, e.g. 5 nines up-time, latency requirements etc
+
+### Multiple Regions
+Some services are served in both the EU and US regions. For services in the US,
+we cannot share any data to the EU. For the services that my teams write, we
+have certain integrations in the EU - Swedish national bank, the European
+Central bank. These of course should not be made available in the US.
+
+### Testability
+We need to be able to test different components in isolation. Some of these
+components are pure or can be testing in a pure manner - if we mock out certain
+dependencies.
+
+### Maintainability
+What is different here than for a regular service? Well, for starters it *is*
+possible to write some really obscure code using a combination of Haskell and
+compiler extensions.
+
+As technical lead for our teams writing Haskell applications, it is my job to
+make sure that others are productive. Thus, this article will emphasize
+*writing easy to grok, maintainable code.*
+
+## Proposed solution
+We will write a configurable, testable service that deploys to a cloud
+provider. The application will be shipped in a docker container and deployable
```

Comment:
Nit-picking, but I do believe it's a Docker image that's shipped and that is then executed in a container.